hash_practice/exercises.py

```diff
@@ -22,8 +48,38 @@ def valid_sudoku(table):
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n^2)
+        Space Complexity: O(n)
     """
-    pass
+    # check if the input has too many rows
+    if len(table) > 9:
+        return False
+    for i in range(len(table)):
+        row_dictionary = {}
+        column_dictionary = {}
+        subgrid_dictionary = {}
+
+        # check if the input has too many columns
+        if len(table[i]) > 9:
+            return False
+        
+        for j in range(len(table[i])):
+            # check row
+            if table[i][j] != "." and table[i][j] in row_dictionary:
+                return False
+            row_dictionary[table[i][j]] = "yay!"
+
+            # check column
+            if table[j][i] != "." and table[j][i] in column_dictionary:
+                return False
+            column_dictionary[table[j][i]] = "yay!"
+
+            # check sub-grid
+            # ????

```

Comment:
As this question is optional, not finishing this is fine, but to give a bit of a hint...

I don't think you are going to really be able to do this in the same loop as the rows and columns. You are going to need another nested loop over the 3x3 sub grid, which is itself nested in a loop over each sub grid. In the instructor solution, the indices of the the subgrids are actually written out as a constant array of pairs, but there's other ways.

But once again, no need to answer this, just giving a hint in case you want to come back and answer this.